lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,659 @@
+  // `entry` bracket
+  revng_assert(Label->kind() != EdgeKind::Invalid);
+  if (Label->kind() == EdgeKind::Fake)
+    return;
+
+  // We marshal the `EdgeDescriptor` object to unwrap the underlying edge, not
+  // wrapped in any `GenericGraph` node, and we invert the direction of the edge
+  // if are traversing it in the inverse direction (wrt the direction of the
+  // original directed edge) during the current exploration
+  using NakedEdgeDescriptor = CycleEquivalenceClass<NodeT>::EdgeDescriptor;
+  NakedEdgeDescriptor NakedEdge;
+  if (not IsInverted) {
+    NakedEdge = NakedEdgeDescriptor({ std::get<0>(E)->getBlock(),
+                                      std::get<1>(E)->getBlock(),
+                                      Label->succNum() });
+  } else {
+    NakedEdge = NakedEdgeDescriptor({ std::get<1>(E)->getBlock(),
+                                      std::get<0>(E)->getBlock(),
+                                      Label->succNum() });
+  }
+
+  size_t EquivalenceClassIndex = SizeTMaxValue;
+  auto MapIt = BracketToClass.find(BD);
+  if (MapIt != BracketToClass.end()) {
+    EquivalenceClassIndex = MapIt->second;
+    revng_assert(CycleEquivalenceClasses.size() > EquivalenceClassIndex);
+    auto &EClass = CycleEquivalenceClasses[EquivalenceClassIndex];
+    EClass.insert(NakedEdge);
+  } else {
+    EquivalenceClassIndex = CycleEquivalenceClasses.size();
+    BracketToClass.emplace(BD, EquivalenceClassIndex);
+    ClassToBracketDescriptor.emplace(EquivalenceClassIndex, BD);
+    auto &EClass = CycleEquivalenceClasses.emplace_back(EquivalenceClassIndex);
+    EClass.insert(NakedEdge);
+  }
+
+  EdgeToCycleEquivalenceClassIDMap.insert(NakedEdge, EquivalenceClassIndex);
+}
+
+template<class GraphT, class GT>
+std::string CycleEquivalenceAnalysis<GraphT, GT>::print() {
+
+  std::string Output;
+
+  // Print the Brackets Analysis results
+  Output += "\nOrdered Bracket Analysis Results:\n";
+  for (auto &EClass : CycleEquivalenceClasses) {
+    Output += EClass.print();
+  }
+
+  // Print the correspondence between the Equivalence Class and the Bracket
+  // Descriptor which gave origin to it
+  Output += "\nClass Bracket Correspondence:\n";
+  for (const auto &[Index, BD] : ClassToBracketDescriptor) {
+    auto &Edge = BD.first;
+    auto &Size = BD.second;
+    Output += std::to_string(Index) + " => " + "("
+              + std::get<0>(Edge)->getName().str() + " <-> "
+              + std::get<1>(Edge)->getName().str() + ","
+              + std::to_string(std::get<2>(BD.first)) + "), "
+              + std::to_string(Size) + "\n";
+  }
+
+  return Output;
+}
+
+template<class GraphT, class GT>
+bool CycleEquivalenceAnalysis<GraphT, GT>::isBackedge(BlockEdgeDescriptor &ED) {
+
+  size_t SourceDFSNum = std::get<0>(ED)->getDFSNum();
+  size_t TargetDFSNum = std::get<1>(ED)->getDFSNum();
+
+  // The equal is very important, or we exclude the self loop backedges
+  return TargetDFSNum <= SourceDFSNum;
```

Comment:
I trust this is the right decision since there's a unit test on self-loops. But it's not clear why. Please improve the comment you say it's `very important` but you don't explain why.

Reply:
I added a more verbose explanation here https://github.com/revng/revng-c/blob/b5fe805f75dde9c9ada264ace247d4b989c4f2d7/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L108.

Basically, the point is that a selfloop source and target nodes DFSNum is equal by defintion. If the equal is left out from the comparison, therefore a selfloop will not satisfy the requirement, and will not be detected as a backedge (which in fact, always is).

Reply:
:+1: